api/v1alpha4/clusterclass_types.go

```diff
@@ -0,0 +1,106 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha4
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=clusterclasses,shortName=cc,scope=Namespaced,categories=cluster-api
+// +kubebuilder:storageversion
+
+// ClusterClass is a template which can be used to create managed topologies.
+type ClusterClass struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec ClusterClassSpec `json:"spec,omitempty"`
+}
+
+// ClusterClassSpec describes the desired state of the ClusterClass.
+type ClusterClassSpec struct {
+	// ControlPlane is a reference to a local struct that holds the details
+	// for provisioning the Control Plane for the Cluster.
+	ControlPlane LocalObjectTemplate `json:"controlPlane,omitempty"`
+
+	// Worker describes the worker nodes for the cluster.
+	// It is a collection of node types which can be used to create
+	// the worker nodes of the cluster.
+	// +optional
+	Worker WorkerClass `json:"worker,omitempty"`
+
+	// Infrastructure is a reference to a provider-specific template that holds
+	// the details for provisioning infrastructure specific cluster
+	// for the underlying provider.
+	// The underlying provider is responsible for the implementation
+	// of the template to an infrastructure cluster.
+	Infrastructure LocalObjectTemplate `json:"infrastructure,omitempty"`
+}
+
+// WorkerClass is a collection of deployment classes.
+type WorkerClass struct {
+	// Deployments is a list of machine deployment types that can be used to create
+	// a set of worker plane nodes.
+	Deployments []DeploymentClass `json:"deployments,omitempty"`
+}
+
+// DeploymentClass serves as a template to define a set of worker nodes of the cluster
+// provisioned using the `ClusterClass`.
+type DeploymentClass struct {
+	// Class denotes a type of worker node present in the cluster
+	Class string `json:"class,omitempty"`
```

Comment:
Let's add some more information here that the class needs to be defined within the ClusterClass' managed deployment sections

Reply:
Not sure I get this comment, this is actually the ClusterClass' managed deployment section...

Reply:
```suggestion
	// Class denotes a type of worker node present in the cluster,
	// this name MUST be unique for this Class and can be referenced
	// in the Cluster to create a managed MachineDeployment.
	Class string `json:"class,omitempty"`
```

Reply:
thanks for clarifying! done